lzy/storage/src/test/java/ai/lzy/storage/StorageClientTest.java

```diff
@@ -0,0 +1,166 @@
+    public void setUp() {
+        int port = GrpcUtils.rollPort();
+        String address = "http://localhost:" + port;
+        storageTestDir = Path.of(System.getProperty("user.dir"), "lzy-storage-tests").toAbsolutePath();
+
+        s3Storage = new S3Mock.Builder().withPort(port).withFileBackend(storageTestDir.toString()).build();
+        s3Storage.start();
+
+        var credentials = new AnonymousAWSCredentials();
+        downloadUploadExecutor = Executors.newFixedThreadPool(5);
+        storageClient = new S3ClientWithTransmitter(address, "", "", downloadUploadExecutor);
+
+        AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
+            .withPathStyleAccessEnabled(true)
+            .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(address, "us-west-2"))
+            .withCredentials(new AWSStaticCredentialsProvider(credentials))
+            .build();
+        s3Client.createBucket(bucket);
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        downloadUploadExecutor.shutdownNow();
+        try {
+            //noinspection ResultOfMethodCallIgnored
+            downloadUploadExecutor.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException ie) {
+            // blank
+        }
+
+        s3Storage.stop();
+        clearDir(storageTestDir);
+    }
+
+    void clearDir(Path path) throws IOException {
+        Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                Files.delete(file);
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                Files.delete(dir);
+                return FileVisitResult.CONTINUE;
+            }
+        });
+    }
+
+    void createSparseFile(String filename) throws IOException {
+        try (var file = new RandomAccessFile(storageTestDir.resolve(filename).toFile(), "rw")) {
+            // file size > 100 MB
+            file.setLength(fileSize);
+        }
+    }
+
+    @Test
+    public void testMultipleReaders() throws IOException, InterruptedException {
+        var files = List.of(
+            "test-file-0", "test-file-1", "test-file-2", "test-file-3", "test-file-4"
+        );
+
+        for (String filename : files) {
+            createSparseFile(filename);
+            storageClient.write(URI.create("s3://" + bucket + "/" + filename), storageTestDir.resolve(filename));
+        }
+        for (String filename : files) {
+            storageClient.read(URI.create("s3://" + bucket + "/" + filename), storageTestDir.resolve(filename));
+        }
+
+        for (String filename : files) {
+            assertEquals(fileSize, Files.size(storageTestDir.resolve(filename)));
+        }
+    }
+
+    @Test
+    @Ignore("Must not halt")
```

Comment:
why ignored?

Reply:
transmitter is not thread safe so concurrent reads should not work